### PR TITLE
perf(mito): scan SSTs and memtables in parallel

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
This PR scans SSTs and memtables in parallel. Scanning each file or memtable is still sequential.

The implementation of parallel scan is quite intuitive: It spawns a scan task for each data source and yields the scanned batch to the merge reader via a channel.

There are two new options for parallel scan.
- scan_parallelism: The default parallelism is `1/4` of CPU cores
- parallel_scan_channel_size: The default channel size is `32`

TSBS results
| query type             | parallelism 1 avg (ms) | parallelism 4 avg (ms) |
| ---                    | ---                | ---                |
| cpu-max-all-1          | 55.80              | 27.89              |
| cpu-max-all-8          | 202.82             | 114.03             |
| double-groupby-1       | 1144.48            | 469.67             |
| double-groupby-5       | 2650.60            | 883.85             |
| double-groupby-all     | 4411.23            | 1589.75            |
| groupby-orderby-limit  | 1507.01            | 502.63             |
| high-cpu-1             | 52.65              | 24.80              |
| high-cpu-all           | 8058.48            | 5484.44            |
| lastpoint              | 14225.00           | 7393.31            |
| single-groupby-1-1-1   | 20.81              | 12.80              |
| single-groupby-1-1-12  | 34.29              | 13.47              |
| single-groupby-1-8-1   | 68.47              | 42.45              |
| single-groupby-5-1-1   | 28.22              | 16.57              |
| single-groupby-5-1-12  | 42.64              | 19.21              |
| single-groupby-5-8-1   | 94.53              | 61.04              |

Larger parallelism might not always help as the total number of SSTs to scan might be less than the parallelism.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
- https://github.com/GreptimeTeam/greptimedb/issues/2806

### CODE REVIEW

File: src/mito2/src/config.rs [hunk 1]
```diff
@@ -68,12 +70,18 @@ pub struct MitoConfig {
     // Other configs:
     /// Buffer size for SST writing.
     pub sst_write_buffer_size: ReadableSize,
+    /// Parallelism to scan a region (default: 1/4 of cpu cores).
+    ///
+    /// 0 or 1 to disable parallelism scan.
+    pub scan_parallelism: usize,
+    /// Capacity of the channel to send data from parallel scan tasks to the main task (default 32).
+    pub parallel_scan_channel_size: usize,
```
Comment (fengjiachun) [inline src/mito2/src/config.rs:78]: Can we just make `scan_parallelism` and `parallel_scan_channel_size` as a linear multiplicative relationship, that is to say, `parallel_scan_channel_size` = n * `scan_parallelism'?

File: src/mito2/src/read/seq_scan.rs [hunk 2]
```diff
@@ -179,16 +227,50 @@ impl SeqScan {
...
+
+    /// Returns whether to use a parallel reader.
+    fn use_parallel_reader(&self) -> bool {
+        self.parallelism.allow_parallel_scan() && (self.files.len() + self.memtables.len()) > 1
+    }
+
+    /// Scan the input source in another task.
+    fn spawn_scan_task(&self, mut input: Source, semaphore: Arc<Semaphore>) -> BoxedBatchStream {
+        let (sender, receiver) = mpsc::channel(self.parallelism.channel_size);
+        tokio::spawn(async move {
```
Comment (WenyXu) [inline src/mito2/src/read/seq_scan.rs:251]: Should we use `comon_runtime`?

File: src/mito2/src/config.rs (hunk 1 above)
Comment (evenyag) [reply src/mito2/src/config.rs:78]: Each parallel scan task creates a channel so they won't share the same channel. Typically the default channel size is enough.

File: src/mito2/src/read/seq_scan.rs (hunk 2 above)
Comment (evenyag) [reply src/mito2/src/read/seq_scan.rs:251]: Yes. I used to use read runtime but I changed it to current runtime finally. I'm not sure whether we should maintain so many runtimes.

Comment (WenyXu) [reply src/mito2/src/read/seq_scan.rs:251]: Some question, too. Does it seem helpful for profiling?

Comment (evenyag) [reply src/mito2/src/read/seq_scan.rs:251]: Not always helpful. We might still need to use pprof.

**Review by WenyXu — APPROVED** [review]

**Review by fengjiachun — APPROVED** [review]
LGTM

**Pull request merged**